libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt

```diff
@@ -616,29 +614,42 @@ internal object MapboxRouteLineUtils {
         return ranges
     }
 
-    private fun getIntersectionsWithGeometryIndex(steps: List<LegStep>?): List<StepIntersection> {
-        return steps
+    private fun getRoadClassArray(
+        steps: List<LegStep>?
+    ): Array<String?> {
+        val intersectionsWithGeometryIndex = steps
             ?.mapNotNull { it.intersections() }
             ?.flatten()
             ?.filter {
                 it.geometryIndex() != null
             }?.toList() ?: listOf()
-    }
 
-    private fun getRoadClassArray(
-        intersectionsWithGeometryIndex: List<StepIntersection>
-    ): Array<String?> {
-        return if (intersectionsWithGeometryIndex.isNotEmpty()) {
-            arrayOfNulls<String>(
-                intersectionsWithGeometryIndex.last().geometryIndex()!! + 1
-            ).apply {
-                intersectionsWithGeometryIndex.forEach {
-                    this[it.geometryIndex()!!] = it.mapboxStreetsV8()?.roadClass()
-                        ?: "intersection_without_class_fallback"
+        return when (intersectionsWithGeometryIndex.isNotEmpty()) {
+            true -> {
+                arrayOfNulls<String>(
+                    intersectionsWithGeometryIndex.last().geometryIndex()!! + 1
+                ).apply {
+                    intersectionsWithGeometryIndex.forEach {
+                        val roadClass: String =
+                            it.mapboxStreetsV8()?.roadClass()
+                                ?: "intersection_without_class_fallback"
+                        if (it.geometryIndex()!! < this.size) {
+                            this[it.geometryIndex()!!] = roadClass
+                        } else {
+                            logE(
+                                TAG,
+                                "Geometry index for step intersection unexpected or" +
+                                    " incorrect. Intersections are expected to have geometry " +
+                                    "indexes in ascending order. There is a risk of incorrect" +
+                                    " congestion or restrictions being shown on the route."
```

Comment:
```suggestion
                                "Geometry index for step intersection unexpected or" +
                                    " incorrect. There is a risk of incorrect" +
                                    " road class styling applied to the route line."
```
Sorry, this is only about road classes and nothing else, isn't it?

Wouldn't we face a similar issues for other annotations though?

Reply:
Right it's just a road classes issue.

Reply:
> Wouldn't we face a similar issue for other annotations though?

There may be data missing from the gradient that gets calculated. This malformed route that was added to the unit test is the one submitted by the external developer.  The unit test sends the route through the function that does all of the processing to extract data about the traffic congestion, road closures, restricted road sections etc.  None of the other functions needed updating or crashed based on the malformed route provided.